ufl/algorithms/replace.py

```diff
@@ -25,11 +26,19 @@ def __init__(self, mapping):
             error("Replacement expressions must have the same shape as what they replace.")
 
     def expr(self, o, *args):
+        if isinstance(o, ExternalOperator):
+            return self._external_operators(o)
+
         try:
             return self.mapping[o]
         except KeyError:
             return self.reuse_if_untouched(o, *args)
 
+    def _external_operators(self, o):
```

Comment:
Indeed, it is better with a public handler. Yes it is because of the nature of the ExternalOperator when replacing a given quantity we need to make sure that we also replace adequately the operands. For instance, this situation happens when replacing the outputs by the saved outputs in pyadjoint.